### ISSUE TITLE
[dev] TP3 Gremlin incompatibilities on TitanDB implementation

### ISSUE THREAD

**Issue by timhemel** [octopus-platform/joern#123]
I've been puzzling to write some TP3 Gremlin traversals on the TitanDB implementation and noticed some things that are different from the Gremlin documentation.

The Gremlin reference manual talks about a Graph step: http://tinkerpop.apache.org/docs/3.1.1-incubating/reference/#graph-step

In theory, you should be able to do:

```
g.V()
.has('functionId','819')
.aggregate('first')
.V()
.has('functionId','884')
.values('functionId')
.where(P.without('first'))
```

In practice, I needed to do this:
```
g.V()
.has('functionId','819')
.aggregate('first')
.getGraph().get().traversal()  // step needed to be able to call V()
.V()
.has('functionId','884')
.values('functionId')
.where(P.without('first'))
```
This however, only takes care of the type errors. The collection `first` is not available after calling the second `V` step, as shown by the error message in the following traversal:
```
g.V()
.has('functionId','819')
.aggregate('first')
.where(P.within('first'))  // this works fine
.getGraph().get().traversal()
.V()                               // this makes 'first' unavailable
.has('functionId','884')
.values('functionId')
.where(P.without('first')) // here we get an error
```
The error is:
```
Neither the sideEffects, map, nor path has a first-key: WherePredicateStep(without([first]))
```
Also note that to be able to create the `where` predicate, I needed to call `P.where`.

I am not sure if this behaviour is a TitanDB specific issue. It limits the way in which we can implement Graph traversals in Joern and may have a performance impact because some traversals cannot have a lazy implementation.

**Comment by a0x77n** [comment]
I think the Graph Step is not available in the version of TP3 we are currently using (3.0.1). 
Maybe we cannot call `where` without the Sugar plugin loaded?

**Comment by timhemel** [comment]
using `P.where` solved it, and is an acceptable workaround IMO. However it would be nicer to be able to use the same syntax as in the documentation.

**Comment by timhemel** [comment]
I was not aware of the syntax differences between Gremlin versions. Is the project planning to support the latest Gremlin syntax, or a limited subset in order to be backwards compatible with older versions of TP3?

**Comment by fabsx00** [comment]
I would say we stick to 3.0.1 incubating, as that is the version included with TitanDB 1.0. I can imagine that if we mix TitanDB with a different TP3 version, many things will start to blow up.
